src/ext/ddtrace.c

```diff
@@ -327,11 +328,64 @@ static PHP_FUNCTION(dd_trace_check_memory_under_limit) {
     }
 }
 
+static PHP_FUNCTION(dd_tracer_circuit_breaker_register_error) {
+    PHP5_UNUSED(return_value_used, this_ptr, return_value_ptr, ht TSRMLS_CC);
+    PHP7_UNUSED(execute_data);
+
+    dd_tracer_circuit_breaker_register_error();
+
+    RETURN_BOOL(1);
+}
+
+static PHP_FUNCTION(dd_tracer_circuit_breaker_register_success) {
+    PHP5_UNUSED(return_value_used, this_ptr, return_value_ptr, ht TSRMLS_CC);
+    PHP7_UNUSED(execute_data);
+
+    dd_tracer_circuit_breaker_register_success();
+
+    RETURN_BOOL(1);
+}
+
+static PHP_FUNCTION(dd_tracer_circuit_breaker_open) {
```

Comment:
As far as I can tell, the only functions that are being used in userland are `dd_tracer_circuit_breaker_can_try()`, `dd_tracer_circuit_breaker_register_error()`, and `dd_tracer_circuit_breaker_register_success()`. So I'm curious if the other functions are just exposing an API for testing purposes? If so, I'm wondering if there's another solution that doesn't require adding additional functions in userland. Maybe only exposing them when compiling in debug mode or something? What are your thoughts? :)

Reply:
I wonder whats the downside keeping them in?
I would prefer to test as close to the final binary as possible so functions only exposed in tests would mean that we have two different binaries - which in turn could foster theoretical edge-cases which wouldn't be detected in tests.

Also those functions provide more access to the Circuit breaker and could be used to extend/ enhance the functioning a bit in the future.

Reply:
> I wonder whats the downside keeping them in?

I'm just thinking of the commitment to ongoing maintenance and not exposing any API's that aren't necessary so we have fewer issues if we need to change or remove them. But I'm not saying we shouldn't add them, I'm just wondering if we have sufficient reason for adding them and are OK with any tradeoffs for adding them. So if you're confident, I'm confident too. :)

> could be used to extend/ enhance the functioning a bit in the future.

I personally don't like adding code for possible future features since the domain could change and we have to commit to maintaining the future code for an indeterminate amount of time. In my experience, most times I've added future code, the future code is never used. :D 

> I would prefer to test as close to the final binary as possible

I agree - I think "to ensure stability" is a pretty good reason to expose these. I would prefer a method to ensure stability that doesn't expose any more public API's, but if there's really no other viable alternative, I'm on board with this! :)

Reply:
Excellent points Sammy!

I think I could remove a few functions though. Will probably be able to drop `_close` and `_open`. Probably `_is_closed` to since its functionality is implemented by a new `_info` .